### PR TITLE
Claims reconn tests

### PR DESCRIPTION
**Pull request by amichal**
Add some unit test coverage to the (surprisingly complex) logic for determining/reporting if a qualifying activity occurred during a valid enrollment and with a valid care plan.

@awisema Take a scan of the `active_patient` cases in `drivers/claims_reporting/spec/reconciliation_report_spec.rb.` I'm fairly confident i got them correct and they match all real world examples I'm aware off. There are a number of cases each described by a block ending in `active_patients << [patient, 1, 0, [cp.provider_signed_on.to_date]]` where the tuple fields are: 
- patient
- enrollment_missing?
- careplan_missing?
- Array of all care plan provider dates for that patient.

We consider a care plan signed if the provider has signed it at the time of the report and we consider it applicable to an enrollment/activity if either the provider or patient signed it at any time during the enrollment and it has not expired by the time of the activity. Effectively, this gives everyone the benefit of the doubt on the exact dates of signature as long as it gets signed around the time of the activity.

### CODE REVIEW

File: app/models/health/qualifying_activity.rb [hunk 1]
```diff
@@ -660,37 +661,50 @@ def missing_care_plan?
...
+      contributing_care_plans = pcp_signed_plans.select do |cp|
+        # Not sure on this... dont penalize the patient if the provider was late signing it
+        cp_date = [cp.provider_signed_on, cp.patient_signed_on].compact.min
+        (
+          (cp.expires_on.nil? || date_of_activity <= cp.expires_on) &&
+          (cp_date >= first_enrollment_date) &&
+          (last_enrollment_date.nil? || cp_date <= last_enrollment_date)
+        )
       end
+      return false if contributing_care_plans.any? && !activity.in?(%w/care_planning pctp_signed/)
```
Comment (awisema) [inline app/models/health/qualifying_activity.rb]: I would prefer the `%w` be replaced with `['care_planning', 'pctp_signed']` -- we have avoided this syntax in OP.

File: drivers/claims_reporting/spec/reconciliation_report_spec.rb [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+             pending_disenrollment_date: month.end_of_month,
+             disenrollment_date: nil)
+      cp = create(:careplan, patient: patient, provider_signed_on: month.beginning_of_month + 15.days)
+      create(:qualifying_activity,
+             patient: patient,
+             date_of_activity: month.beginning_of_month + 11.days,
+             claim_submitted_on: claim_submitted_on)
+      active_patients << [patient, 0, 0, [cp.provider_signed_on.to_date]]
+    end
+    # 7. after pending disenrollment before actual disenrollment, after valid careplan
```
Comment (awisema) [inline drivers/claims_reporting/spec/reconciliation_report_spec.rb:107]: This points to a possible issue in the current codebase... I think we should be able to enforce that if there is a `disenrollment_date`, the 'pending_disenrollment_date` is cleared...

File: app/models/health/qualifying_activity.rb (hunk 1 above)
Comment (amichal) [reply app/models/health/qualifying_activity.rb]: Got it, Will push a commit with rubocop -A on that file to tidy it all up

File: drivers/claims_reporting/spec/reconciliation_report_spec.rb (hunk 2 above)
Comment (amichal) [reply drivers/claims_reporting/spec/reconciliation_report_spec.rb:107]: Agreed but there is no current validation so wanted to ensure that we did the right thing with messy data here. A separate PR to add the validation and clean up existing data perhaps?

File: app/models/health/qualifying_activity.rb [hunk 3]
```diff
@@ -231,22 +231,22 @@ def self.activities
...
     def self.date_search(start_date, end_date)
       if start_date.present? && end_date.present?
-        self.where("date_of_activity >= ? AND date_of_activity <= ?", start_date, end_date)
+        where('date_of_activity >= ? AND date_of_activity <= ?', start_date, end_date)
       elsif start_date.present?
-        self.where("date_of_activity >= ?", start_date)
+        where('date_of_activity >= ?', start_date)
       elsif end_date.present?
-        self.where("date_of_activity <= ?", end_date)
+        where('date_of_activity <= ?', end_date)
```
Comment (eanders) [inline app/models/health/qualifying_activity.rb:243]: These should be switched over to Arel in the future.

**Pull request merged**